mlflow/pyfunc/__init__.py

```diff
@@ -1096,6 +1152,38 @@ def _predict_row_batch(predict_fn, args):
             StringType: np.str_,
         }
 
+        def _convert_array_values(values, spark_type):
+            array_dim = 1
+            elem_type = spark_type.elementType
+            if isinstance(elem_type, ArrayType):
+                elem_type = elem_type.elementType
+                if isinstance(elem_type, ArrayType):
+                    raise MlflowException(
+                        f"Unsupported array<array<array<ElementType>>> type field "
+                        "in struct type."
+                    )
+                array_dim = 2
+
+            if type(elem_type) not in spark_primitive_type_to_np_type:
+                raise MlflowException(
+                    "Unsupported array type field with element type "
+                    f"{elem_type.simpleString()} in struct type.",
+                    error_code=INVALID_PARAMETER_VALUE,
+                )
+            np_type = spark_primitive_type_to_np_type[type(elem_type)]
+
+            if array_dim == 1:
+                return [
+                    np.array(v, dtype=np_type) for v in values
+                ]
+            else:
+                return [
+                    list(np.array(v, dtype=np_type)) for v in values
```

Comment:
can we use `tolist()`?

Reply:
What happens if we `tolist()`?

Reply:
Here `v` is a 2D array, if we call `tolist`, it converts nested array to list as well, it is less performant.

>>> v1 = np.array([[2,3], [4,5]])
>>> v1.tolist()
[[2, 3], [4, 5]]
>>> list(v1)
[array([2, 3]), array([4, 5])]

Reply:
Got it. Whatever works fine with pyspark is fine.